security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -114,6 +121,54 @@ func (sf *SecretFetcher) Run(ch chan struct{}) {
 	cache.WaitForCacheSync(ch, sf.scrtController.HasSynced)
 }
 
+// Generate implements the XDS Generate method for SDS.
+// We also allow returning the secrets over the main XDS connection,
+// for efficiency.
+// This method only allows returning secret for authenticated requests.
+func (sf *SecretFetcher) Generate(proxy *istiomodel.Proxy, push *istiomodel.PushContext,
+	w *istiomodel.WatchedResource, updates istiomodel.XdsUpdates) istiomodel.Resources {
+	// 1. Get the namespace of the caller
+	// TODO: add an error to Generate (for non-empty)
+
+	// This depends on in-progress PR to VERIFY the namespace.
+
+	// 2. If istio-system - ok, can read secrets in istio-system
+	// This istiod can either be the 'master' in istio-system, or
+	// an replica of istiod running in a namespace with reduced
+	// RBAC to only namespace secret access. The second is equivalent
+	// with running the gateway with secret access, but secret permissions
+	// are granted to a separate istiod pod.
+	if proxy.Metadata.Namespace != sf.secretNamespace {
+		// First iteration: we only support istio-system ( or ingress + istiod in same namespace )
+		// If Ingress is in a different namespace - either use the old code or
+		// run a local istiod for SDS access.
+		return nil
+	}
+
+
+	// 3. Else: it must be in mesh config
+	// TODO: if mesh config for this SA exists, allow access
+	// Also remove the creation of Secrets.
+	// This is for the 'Istiod minting DNS certs signed with
+	//  citadel for other Istiod replicas', in central istiod
+
+	// Get the secret, if any
+	si, f := sf.secrets.Load(w.ResourceNames[0])
```

Comment:
Do we really need the secretsfetcher complexity? This really could be reduced to pretty much 1 line: `secretsLister.Get(w.ResourceNames[0])`.

Perfectly fine with using the existing code in short term, but long term do we need it?

Reply:
Good point - the reason I kept it was because the interface may have multiple implementations. Now we read only from K8S secrets - but one of the goals of this change is to allow integration with external secret stores. 

And just like cert signing, there are 2 integration points - Agent using the SDS implementation of an external store that fully understand Istio Spifee and policies. Or Istiod doing the auth/checks, and using a secret fetcher implementation that calls the secret store with istiod's credentials.

Reply:
Agreed on abstraction to have different implementations. But that doesn't mean we need to make the k8s one more complicated, we can have both - a generic interfaced backed by `secretsLister.Get(w.ResourceNames[0])`

Reply:
I see what you mean: replace the entire convoluted implementation with 3-4 lines of code. We no longer need the 'cache' ( used in the old SDS as a roundabout way to implement ACK/NACK ), nor do we need almost anything else in the class. 

I am planning to do that - but got bit by 'PR is too big, please split it' :-)

Reply:
Yep, 100% fine with doing it in another PR, just something we want to do eventually.